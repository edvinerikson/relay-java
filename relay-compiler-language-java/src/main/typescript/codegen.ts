import {
  ReaderFragment,
  ReaderFragmentSpread,
  ReaderSelection,
  ReaderLinkedField,
  ReaderScalarField,
  ReaderCondition,
  ReaderInlineFragment,
  ReaderInlineDataFragmentSpread,
  ReaderInlineDataFragment
} from "relay-runtime/lib/util/ReaderNode";
import { GeneratedNode, ConcreteRequest } from "relay-runtime";
import invariant from "invariant";

function str(text: string): string {
  return JSON.stringify(text);
}
/*
function printArgumentDefinition(node: ReaderArgumentDefinition): string {
  switch (node.kind) {
    case "LocalArgument": {
      node = node as ReaderLocalArgument;
      console.log(node);
      return `
        LocalArgument
          .builder()

      `;
    }
    case "RootArgument": {
      node = node as ReaderRootArgument;
    }

    default: {
      invariant(
        false,
        'Unepected kind "%s". This is a bug in relay-compiler-language-java',
        node.kind
      );
    }
  }
}

function printArgumentDefinitions(
  argumentDefinitions:
    | ReaderArgumentDefinition[]
    | readonly ReaderArgumentDefinition[]
): string {
  return `Arrays.asList(${(argumentDefinitions as ReaderArgumentDefinition[])
    .map(printArgumentDefinition)
    .join(", ")})`;
}

function printArgument(node: ReaderArgument) {
  switch (node.kind) {
    case "Literal": {
      let field = node as ReaderLiteral;
        return `
        Literal
          .builder()
          .name(${str(field.name)})
          .value(${
            typeof field.value === "string" ? str(field.value)
              : field.value
          })
          .build()
      `;
    }
    case "Variable": {
      let field = node as ReaderVariable;
      return `
        Variable
          .builder()
          .name(${str(field.name)})
          .variableName(${str(field.variableName)})
          .build()
      `;
    }
    default: {
      invariant(
        false,
        'Unepected kind "%s". This is a bug in relay-compiler-language-java',
        node.kind
      );
    }
  }
}

function printArguments(
  nodes: ReaderArgument[] | readonly ReaderArgument[] | null | undefined
) {
  return nodes == null
    ? "Arrays.asList()"
    : `Arrays.asList(${(nodes as ReaderArgument[])
        .map(printArgument)
        .join(", ")})`;
}
*/

function printSelection(node: ReaderSelection): string {
  switch (node.kind) {
    case "LinkedField": {
      let field = node as ReaderLinkedField;
      return `
        LinkedField
          .builder()
          .name(${str(field.name)})
          .alias(${str(field.alias ?? field.name)})
          .selections(${printSelections(field.selections)})
          .build()
      `;
    }

    case "ScalarField": {
      let field = node as ReaderScalarField;
      return `
        ScalarField
          .builder()
          .name(${str(field.name)})
          .alias(${str(field.alias ?? field.name)})
          .build()
      `;
    }

    case "Condition": {
      let field = node as ReaderCondition;
      return `
        Condition
          .builder()
          .condition(${str(field.condition)})
          .passingValue(${field.passingValue})
          .selections(${printSelections(field.selections)})
          .build()
      `;
    }

    case "InlineFragment": {
      let field = node as ReaderInlineFragment;
      return `
        InlineFragment
          .builder()
          .type(${str(field.type)})
          .selections(${printSelections(field.selections)})
          .build()
      `;
    }

    case "FragmentSpread": {
      let field = node as ReaderFragmentSpread;
      return `
        FragmentSpread
          .builder()
          .name(${str(field.name)})
          .build()
      `;
    }

    case "InlineDataFragmentSpread": {
      let field = node as ReaderInlineDataFragmentSpread;

      return `
        InlineDataFragmentSpread
          .builder()
          .name(${str(field.name)})
          .selections(${printSelections(field.selections)})
          .build()
      `;
    }

    default: {
      invariant(
        false,
        'Unepected kind "%s". This is a bug in relay-compiler-language-java',
        node.kind
      );
    }
  }
}

function printSelections(
  selections: readonly ReaderSelection[] | ReaderSelection[]
) {
  return `Arrays.asList(${(selections as ReaderSelection[])
    .map(printSelection)
    .join(", ")})`;
}

/**
 * Converts Relay IR AST to Java code
 * @param node
 */
export function print(
  node: GeneratedNode | ReaderFragmentSpread | ReaderFragment
): string {
  switch (node.kind) {
    case "InlineDataFragment": {
      node = node as ReaderInlineDataFragment;
      return `
        public class ${node.name} extends InlineDataFragment {
          public ${node.name}() {
            super(${str(node.name)});
          }
        }
      `;
    }
    case "Fragment": {
      node = node as ReaderFragment;
      return `
        public class ${node.name} extends Fragment {
          public ${node.name}() {
            super(
              ${str(node.name)},
              ${node.metadata?.mask ?? true},
              ${node.metadata?.plural ?? false},
              ${printSelections(node.selections)}
            );
          }
        }
      `;
    }

    case "Request": {
      node = node as ConcreteRequest;
      return `
        public class ${node.params.name} extends Request {
          public ${node.params.name}() {
            super(
              ${str(node.params.name)},
              Fragment
                .builder()
                .name(${str(node.fragment.name)})
                .masked(${node.fragment.metadata?.mask ?? true})
                .plural(${node.fragment.metadata?.plural ?? false})
                .selections(${printSelections(node.fragment.selections)})
                .build()
            );
          }
        }
      `;
    }
    default: {
      invariant(
        false,
        'Got unexpected "%s". This is likely a bug in relay-compiler-language-java.',
        node.kind
      );
    }
  }
}
