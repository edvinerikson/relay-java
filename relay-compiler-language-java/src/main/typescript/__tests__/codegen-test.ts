import { print } from "../codegen";

import { CompilerContext, IRTransforms } from "relay-compiler";

import {
  parseGraphQLText,
  generateTestsFromFixtures,
  TestSchema,
  compileArtifacts
} from "../test-utils/testUtils";
import { format } from "prettier";
describe("print(node)", () => {
  generateTestsFromFixtures(`${__dirname}/fixtures/print`, text => {
    const relaySchema = TestSchema.extend(IRTransforms.schemaExtensions);
    const { definitions, schema } = parseGraphQLText(relaySchema, text);
    const compilerContext = new CompilerContext(schema).addAll(definitions);

    compilerContext.applyTransforms([
      ...IRTransforms.commonTransforms,
      ...IRTransforms.queryTransforms
    ]);
    return compileArtifacts(compilerContext)
      .map(node => {
        let out = "";
        out += print(node) + "\n\n";

        try {
          return format(out, {
            filepath: "test.java",
            plugins: [require("prettier-plugin-java")]
          });
        } catch (err) {
          out += err.stack;
          return out;
        }
      })
      .join("\n\n");
  });
});
