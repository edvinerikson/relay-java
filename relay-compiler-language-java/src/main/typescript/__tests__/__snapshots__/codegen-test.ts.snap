// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`print(node) matches expected output: basic-query.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
query GetCurrentBackOfficeAgentQuery {
  currentBackofficeAgent {
    __typename
    id
    name
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
public class GetCurrentBackOfficeAgentQuery extends Request {

  public GetCurrentBackOfficeAgentQuery() {
    super(
      "GetCurrentBackOfficeAgentQuery",
      Fragment
        .builder()
        .name("GetCurrentBackOfficeAgentQuery")
        .masked(true)
        .plural(false)
        .selections(
          Arrays.asList(
            LinkedField
              .builder()
              .name("currentBackofficeAgent")
              .alias("currentBackofficeAgent")
              .selections(
                Arrays.asList(
                  ScalarField
                    .builder()
                    .name("__typename")
                    .alias("__typename")
                    .build(),
                  ScalarField.builder().name("id").alias("id").build(),
                  ScalarField.builder().name("name").alias("name").build()
                )
              )
              .build()
          )
        )
        .build()
    );
  }
}

`;

exports[`print(node) matches expected output: query-with-alias.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
query AliasedQuery {
  agent: currentBackofficeAgent {
    name
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
public class AliasedQuery extends Request {

  public AliasedQuery() {
    super(
      "AliasedQuery",
      Fragment
        .builder()
        .name("AliasedQuery")
        .masked(true)
        .plural(false)
        .selections(
          Arrays.asList(
            LinkedField
              .builder()
              .name("currentBackofficeAgent")
              .alias("agent")
              .selections(
                Arrays.asList(
                  ScalarField.builder().name("name").alias("name").build()
                )
              )
              .build()
          )
        )
        .build()
    );
  }
}

`;

exports[`print(node) matches expected output: query-with-arguments.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
query GetStaticPlayerQuery {
  player(playerUid: "123456789") {
    __typename
  }
}

query GetAllPlayersQuery {
  allPlayers(includeDeleted: true) {
    __typename
  }
}

query GetDeletedPostsQuery {
  posts(filter: { deleted: true }) {
    title
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
public class GetStaticPlayerQuery extends Request {

  public GetStaticPlayerQuery() {
    super(
      "GetStaticPlayerQuery",
      Fragment
        .builder()
        .name("GetStaticPlayerQuery")
        .masked(true)
        .plural(false)
        .selections(
          Arrays.asList(
            LinkedField
              .builder()
              .name("player")
              .alias("player")
              .selections(
                Arrays.asList(
                  ScalarField
                    .builder()
                    .name("__typename")
                    .alias("__typename")
                    .build()
                )
              )
              .build()
          )
        )
        .build()
    );
  }
}


public class GetAllPlayersQuery extends Request {

  public GetAllPlayersQuery() {
    super(
      "GetAllPlayersQuery",
      Fragment
        .builder()
        .name("GetAllPlayersQuery")
        .masked(true)
        .plural(false)
        .selections(
          Arrays.asList(
            LinkedField
              .builder()
              .name("allPlayers")
              .alias("allPlayers")
              .selections(
                Arrays.asList(
                  ScalarField
                    .builder()
                    .name("__typename")
                    .alias("__typename")
                    .build()
                )
              )
              .build()
          )
        )
        .build()
    );
  }
}


public class GetDeletedPostsQuery extends Request {

  public GetDeletedPostsQuery() {
    super(
      "GetDeletedPostsQuery",
      Fragment
        .builder()
        .name("GetDeletedPostsQuery")
        .masked(true)
        .plural(false)
        .selections(
          Arrays.asList(
            LinkedField
              .builder()
              .name("posts")
              .alias("posts")
              .selections(
                Arrays.asList(
                  ScalarField.builder().name("title").alias("title").build()
                )
              )
              .build()
          )
        )
        .build()
    );
  }
}

`;

exports[`print(node) matches expected output: query-with-conditional.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
query ConditionalQuery {
  player {
    __typename
  }
  currentBackofficeAgent @include(if: false) {
    name
  }
}

query ConditionalVariableQuery($enabled: Boolean!) {
  currentBackofficeAgent @include(if: $enabled) {
    name
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
public class ConditionalQuery extends Request {

  public ConditionalQuery() {
    super(
      "ConditionalQuery",
      Fragment
        .builder()
        .name("ConditionalQuery")
        .masked(true)
        .plural(false)
        .selections(
          Arrays.asList(
            LinkedField
              .builder()
              .name("player")
              .alias("player")
              .selections(
                Arrays.asList(
                  ScalarField
                    .builder()
                    .name("__typename")
                    .alias("__typename")
                    .build()
                )
              )
              .build()
          )
        )
        .build()
    );
  }
}


public class ConditionalVariableQuery extends Request {

  public ConditionalVariableQuery() {
    super(
      "ConditionalVariableQuery",
      Fragment
        .builder()
        .name("ConditionalVariableQuery")
        .masked(true)
        .plural(false)
        .selections(
          Arrays.asList(
            Condition
              .builder()
              .condition("enabled")
              .passingValue(true)
              .selections(
                Arrays.asList(
                  LinkedField
                    .builder()
                    .name("currentBackofficeAgent")
                    .alias("currentBackofficeAgent")
                    .selections(
                      Arrays.asList(
                        ScalarField.builder().name("name").alias("name").build()
                      )
                    )
                    .build()
                )
              )
              .build()
          )
        )
        .build()
    );
  }
}

`;

exports[`print(node) matches expected output: query-with-fragments.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
query MyQuery {
  ...query_frag
  ... on Query {
    currentBackofficeAgent {
      name
      ...AgentData
    }
  }
}

fragment query_frag on Query @inline {
  __typename
}
fragment AgentData on BackofficeAgent {
  id
  name
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
public class MyQuery extends Request {

  public MyQuery() {
    super(
      "MyQuery",
      Fragment
        .builder()
        .name("MyQuery")
        .masked(true)
        .plural(false)
        .selections(
          Arrays.asList(
            LinkedField
              .builder()
              .name("currentBackofficeAgent")
              .alias("currentBackofficeAgent")
              .selections(
                Arrays.asList(
                  ScalarField.builder().name("name").alias("name").build(),
                  FragmentSpread.builder().name("AgentData").build()
                )
              )
              .build(),
            InlineDataFragmentSpread
              .builder()
              .name("query_frag")
              .selections(
                Arrays.asList(
                  ScalarField
                    .builder()
                    .name("__typename")
                    .alias("__typename")
                    .build()
                )
              )
              .build()
          )
        )
        .build()
    );
  }
}


public class query_frag extends InlineDataFragment {

  public query_frag() {
    super("query_frag");
  }
}


public class AgentData extends Fragment {

  public AgentData() {
    super(
      "AgentData",
      true,
      false,
      Arrays.asList(
        ScalarField.builder().name("id").alias("id").build(),
        ScalarField.builder().name("name").alias("name").build()
      )
    );
  }
}

`;

exports[`print(node) matches expected output: query-with-union.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
query PlayerResultInlineFragmentQuery {
  player(playerUid: "123") {
    __typename
    ... on Player {
      username
    }
    ... on PlayerNotFound {
      message
    }
  }
}

query PlayerResultFragmentQuery {
  player(playerUid: "123") {
    __typename
    ...PlayerFragment
    ...PlayerNotFound
  }
}

fragment PlayerFragment on Player {
  username
}

fragment PlayerNotFound on PlayerNotFound {
  message
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
public class PlayerResultInlineFragmentQuery extends Request {

  public PlayerResultInlineFragmentQuery() {
    super(
      "PlayerResultInlineFragmentQuery",
      Fragment
        .builder()
        .name("PlayerResultInlineFragmentQuery")
        .masked(true)
        .plural(false)
        .selections(
          Arrays.asList(
            LinkedField
              .builder()
              .name("player")
              .alias("player")
              .selections(
                Arrays.asList(
                  ScalarField
                    .builder()
                    .name("__typename")
                    .alias("__typename")
                    .build(),
                  InlineFragment
                    .builder()
                    .type("Player")
                    .selections(
                      Arrays.asList(
                        ScalarField
                          .builder()
                          .name("username")
                          .alias("username")
                          .build()
                      )
                    )
                    .build(),
                  InlineFragment
                    .builder()
                    .type("PlayerNotFound")
                    .selections(
                      Arrays.asList(
                        ScalarField
                          .builder()
                          .name("message")
                          .alias("message")
                          .build()
                      )
                    )
                    .build()
                )
              )
              .build()
          )
        )
        .build()
    );
  }
}


public class PlayerResultFragmentQuery extends Request {

  public PlayerResultFragmentQuery() {
    super(
      "PlayerResultFragmentQuery",
      Fragment
        .builder()
        .name("PlayerResultFragmentQuery")
        .masked(true)
        .plural(false)
        .selections(
          Arrays.asList(
            LinkedField
              .builder()
              .name("player")
              .alias("player")
              .selections(
                Arrays.asList(
                  ScalarField
                    .builder()
                    .name("__typename")
                    .alias("__typename")
                    .build(),
                  FragmentSpread.builder().name("PlayerFragment").build(),
                  FragmentSpread.builder().name("PlayerNotFound").build()
                )
              )
              .build()
          )
        )
        .build()
    );
  }
}


public class PlayerFragment extends Fragment {

  public PlayerFragment() {
    super(
      "PlayerFragment",
      true,
      false,
      Arrays.asList(
        ScalarField.builder().name("username").alias("username").build()
      )
    );
  }
}


public class PlayerNotFound extends Fragment {

  public PlayerNotFound() {
    super(
      "PlayerNotFound",
      true,
      false,
      Arrays.asList(
        ScalarField.builder().name("message").alias("message").build()
      )
    );
  }
}

`;

exports[`print(node) matches expected output: query-with-variables.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
query GetPlayerQuery($playerId: String) {
  player(playerUid: $playerId) {
    __typename
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
public class GetPlayerQuery extends Request {

  public GetPlayerQuery() {
    super(
      "GetPlayerQuery",
      Fragment
        .builder()
        .name("GetPlayerQuery")
        .masked(true)
        .plural(false)
        .selections(
          Arrays.asList(
            LinkedField
              .builder()
              .name("player")
              .alias("player")
              .selections(
                Arrays.asList(
                  ScalarField
                    .builder()
                    .name("__typename")
                    .alias("__typename")
                    .build()
                )
              )
              .build()
          )
        )
        .build()
    );
  }
}

`;

exports[`print(node) matches expected output: relay-directive.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
fragment MyMaskedFragment on Query {
  currentBackofficeAgent {
    id
  }
}

fragment MyUnmaskedFragment on Query {
  currentBackofficeAgent {
    name
  }
}
query WithRelayDirectiveQuery {
  ...MyUnmaskedFragment @relay(mask: false)
  ...MyMaskedFragment @relay(mask: true)
  ...MyAlwaysMaskedFragment
  ...MyAlwaysUnmaskedFragment

  posts {
    ...MyPluralFragment
  }
}

fragment MyAlwaysMaskedFragment on Query @relay(mask: true) {
  currentBackofficeAgent {
    name
  }
}
fragment MyAlwaysUnmaskedFragment on Query @relay(mask: false) {
  currentBackofficeAgent {
    name
  }
}

fragment MyPluralFragment on Post @relay(plural: true) {
  title
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
public class WithRelayDirectiveQuery extends Request {

  public WithRelayDirectiveQuery() {
    super(
      "WithRelayDirectiveQuery",
      Fragment
        .builder()
        .name("WithRelayDirectiveQuery")
        .masked(true)
        .plural(false)
        .selections(
          Arrays.asList(
            LinkedField
              .builder()
              .name("currentBackofficeAgent")
              .alias("currentBackofficeAgent")
              .selections(
                Arrays.asList(
                  ScalarField.builder().name("name").alias("name").build()
                )
              )
              .build(),
            LinkedField
              .builder()
              .name("posts")
              .alias("posts")
              .selections(
                Arrays.asList(
                  FragmentSpread.builder().name("MyPluralFragment").build()
                )
              )
              .build(),
            FragmentSpread.builder().name("MyMaskedFragment").build(),
            FragmentSpread.builder().name("MyAlwaysMaskedFragment").build(),
            FragmentSpread.builder().name("MyAlwaysUnmaskedFragment").build()
          )
        )
        .build()
    );
  }
}


public class MyMaskedFragment extends Fragment {

  public MyMaskedFragment() {
    super(
      "MyMaskedFragment",
      true,
      false,
      Arrays.asList(
        LinkedField
          .builder()
          .name("currentBackofficeAgent")
          .alias("currentBackofficeAgent")
          .selections(
            Arrays.asList(ScalarField.builder().name("id").alias("id").build())
          )
          .build()
      )
    );
  }
}


public class MyAlwaysMaskedFragment extends Fragment {

  public MyAlwaysMaskedFragment() {
    super(
      "MyAlwaysMaskedFragment",
      true,
      false,
      Arrays.asList(
        LinkedField
          .builder()
          .name("currentBackofficeAgent")
          .alias("currentBackofficeAgent")
          .selections(
            Arrays.asList(
              ScalarField.builder().name("name").alias("name").build()
            )
          )
          .build()
      )
    );
  }
}


public class MyAlwaysUnmaskedFragment extends Fragment {

  public MyAlwaysUnmaskedFragment() {
    super(
      "MyAlwaysUnmaskedFragment",
      false,
      false,
      Arrays.asList(
        LinkedField
          .builder()
          .name("currentBackofficeAgent")
          .alias("currentBackofficeAgent")
          .selections(
            Arrays.asList(
              ScalarField.builder().name("name").alias("name").build()
            )
          )
          .build()
      )
    );
  }
}


public class MyPluralFragment extends Fragment {

  public MyPluralFragment() {
    super(
      "MyPluralFragment",
      true,
      true,
      Arrays.asList(ScalarField.builder().name("title").alias("title").build())
    );
  }
}

`;
